include/mbedtls/ssl.h

```diff
@@ -2529,6 +2529,7 @@ static inline uintptr_t mbedtls_ssl_get_user_data_n(
 }
 
 #if defined(MBEDTLS_SSL_ASYNC_PRIVATE)
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
```

Comment:
1. Async callbacks don't need X.509. Right now they aren't useful without X.509, but when we finally add support for raw public keys, async callbacks will apply to raw public keys. Async callbacks apply to all key exchanges that use either decryption or signature, and they're only implemented on the server side. So the condition should be something like
```
defined(MBEDTLS_SSL_SRV_C) && \
( defined(MBEDTLS_KEY_EXCHANGE_RSA_PSK_ENABLED) || \
  defined(MBEDTLS_KEY_EXCHANGE_RSA_ENABLED) || \
  defined(MBEDTLS_KEY_EXCHANGE_WITH_SERVER_SIGNATURE_ENABLED) )
```
Plus whatever it takes for TLS 1.3.

2. If `MBEDTLS_SSL_ASYNC_PRIVATE` is enabled, this function is always needed. So you're putting the check in the wrong place: it should be a condition for enabling `MBEDTLS_SSL_ASYNC_PRIVATE`, tested in `check_config.h`. But why is this condition needed at all? If the build doesn't include any key exchange for which async private key callbacks can be used, it's still ok if the application sets up the callbacks — they just won't be used.

Reply:
If we want only `crypto_full minus MD` config then this change can be dropped. But for me it seems there is a problem here.

This was added as `mbedtls_ssl_async_sign_t` and `mbedtls_ssl_async_decrypt_t` types used as callback arguments are only available if `MBEDTLS_SSL_ASYNC_PRIVATE` and  `MBEDTLS_X509_CRT_PARSE_C` symbols are defined.

`MBEDTLS_X509_CRT_PARSE_C` is available in case of `crypto_full minus MD`, but not in `full minus MD`.